README.md
Assetto Server Manager (ACSM)
=============================

An advanced web interface for managing Assetto Corsa servers.

## Installation

Hi! Thanks for purchasing ACSM. It means a lot to us, and we hope you enjoy using it!

Let's get started with the installation process. Please note, we don't cover all configurable options here, but these steps
should be enough to get you set up with basic server manager install.

### Single Server

"Single Server" means that you only want to run 1 Assetto Corsa Server. This is the simplest install,
and definitely the best place to start!

#### Windows

1. Extract the zip file you downloaded (you probably already did this, as you're reading this guide!)
2. Make a new folder somewhere you would like to store your Assetto Corsa server. Some sensible places include:
   * C:\Users\YourUsernameHere\AssettoCorsaServer
   * C:\Users\YourUsernameHere\Desktop\AssettoCorsaServer
   * C:\AssettoCorsaServer
   Or somewhere else, it doesn't matter too much! We'll refer to this as your 'server manager install folder' 
   from now on.
3. Take the 'server-manager.exe' and 'config.yml' (the ones you just extracted), 
   and move them in to your server manager install folder.
4. Copy your ACSM.License (you can get this from emperorservers.com, or your purchase email!) 
   into the server manager install folder too. This should now be in the same folder as the server-manager.exe.
5. Open the config.yml to edit it. You can open it in Notepad, or something like NotePad++ or Visual Studio code.
6. There are lots of options in the config.yml, they all have descriptions of what they do. Most are set to 
   sensible defaults - so you don't need to change them all! However, there are a few important ones...
7. Find the "steam" -> "install_path" option. 
   This option tells Server Manager where the Assetto Corsa server is installed.
   You have a few options here:
   a) Set this option to the location where the server is installed, for example:
        "C:\Program Files\Steam\steamapps\common\Assetto Corsa Dedicated Server\"
   b) In explorer, copy the Assetto Corsa server folder into your server manager install folder 
      and rename it to 'assetto', and leave this option set to 'assetto'
   c) Leave this option unchanged. Don't copy any files. Server Manager will create a 'barebones' install
      with the default cars, and you can upload the tracks and more car details later.
   d) Install steamcmd, and set the steam username and password, then let Server Manager install the
      Assetto Corsa server for you.
8. Find the "http" -> "session_key" option. You must change this to a random string. 
   Think of it like a password. Create something random and don't share it with anyone.
9. Save the config.yml
10. You should now be ready to launch server manager. You can do this by double clicking the "server-manager.exe".
11. When it has opened, Server Manager will open your browser. If it doesn't, you can open it by going to
    http://127.0.0.1:8772 in your browser.
12. Log in using the username 'admin' and the password 'servermanager'. You will then be asked to change your
    password.
13. We recommend that you then go to the Tracks page, and follow the instructions to upload your tracks.
14. We also recommend that you go to the Cars page, and follow the instructions to upload your cars. You may need to do
    them in smaller batches, of say 20-30 cars. You can upload multiple at once by selecting them together in windows 
    explorer and dragging them into the Drag and Drop area of the upload form.
15. Go to the Server Options page. Please read through the options carefully. Pay specific attention to the port values of:
    UDP Port, HTTP Port, TCP Port, Content Manager Wrapper Port. These will all need to be made accessible on the web
    by a technique called "Port Forwarding". This is not covered in this guide, but you will find many guides on Google.
    Make sure to save your server options.
16. You can now try running an event! We recommend a 'Quick Race' to test that the server is working OK. Go to the
    Quick Race page, and set up an event. Press 'Start Race'.
17. If you can't join your quick race, or the server goes offline, please go to the 'Server Logs' page, and read the logs
    for both the "Assetto Corsa Server" and "Server Manager". Often errors will be shown in red there.
18. Enjoy using Server Manager!

If you struggle with any of these steps, or Server Manager doesn't open properly, you can check your server-manager.log
for more information on what might have gone wrong!

#### Linux

1. Extract the zip file you downloaded (you probably already did this, as you're reading this guide!)
2. Make a new folder somewhere you would like to store your Assetto Corsa server. An example of a sensible place is:
   * /home/<yourusername/AssettoCorsaServer
   Or somewhere else, it doesn't matter too much! We'll refer to this as your 'server manager install folder' 
   from now on.
3. Take the 'server-manager' binary and 'config.yml' (the ones you just extracted), 
   and move them in to your server manager install folder.
4. Copy your ACSM.License (you can get this from emperorservers.com, or your purchase email!) 
   into the server manager install folder too. This should now be in the same folder as the server-manager.
5. Open the config.yml to edit it. You can open it in Notepad, or something like NotePad++ or Visual Studio code.
6. There are lots of options in the config.yml, they all have descriptions of what they do. Most are set to 
   sensible defaults - so you don't need to change them all! However, there are a few important ones...
7. Find the "steam" -> "install_path" option. 
   This option tells Server Manager where the Assetto Corsa server is installed.
   You have a few options here:
   a) Set this option to the location where the server is installed, for example:
        "/var/lib/steamcmd/Assetto Corsa Server"
   b) Copy the Assetto Corsa server folder into your server manager install folder 
      and rename it to 'assetto', and leave this option set to 'assetto'
   c) Leave this option unchanged. Don't copy any files. Server Manager will create a 'barebones' install
      with the default cars, and you can upload the tracks and more car details later.
   d) Install steamcmd, and set the steam username and password, then let Server Manager install the
      Assetto Corsa server for you.
8. Find the "http" -> "session_key" option. You must change this to a random string. 
   Think of it like a password. Create something random and don't share it with anyone.
9. Save the config.yml
10. You should now be ready to launch server manager. You can do this by running "./server-manager" in terminal from the 
    server manager install path.
11. Open the server manager web UI by going to http://127.0.0.1:8772 (this may be different if you changed the http port!)
12. Log in using the username 'admin' and the password 'servermanager'. You will then be asked to change your
    password.
13. We recommend that you then go to the Tracks page, and follow the instructions to upload your tracks.
14. We also recommend that you go to the Cars page, and follow the instructions to upload your cars. You may need to do
    them in smaller batches, of say 20-30 cars. You can upload multiple at once by selecting them together in windows 
    explorer and dragging them into the Drag and Drop area of the upload form.
15. Go to the Server Options page. Please read through the options carefully. Pay specific attention to the port values of:
    UDP Port, HTTP Port, TCP Port, Content Manager Wrapper Port. These will all need to be made accessible on the web
    by a technique called "Port Forwarding". This is not covered in this guide, but you will find many guides on Google.
    Make sure to save your server options.
16. You can now try running an event! We recommend a 'Quick Race' to test that the server is working OK. Go to the
    Quick Race page, and set up an event. Press 'Start Race'.
17. If you can't join your quick race, or the server goes offline, please go to the 'Server Logs' page, and read the logs
    for both the "Assetto Corsa Server" and "Server Manager". Often errors will be shown in red there.
18. Enjoy using Server Manager!

If you struggle with any of these steps, or Server Manager doesn't open properly, you can check your server-manager.log
for more information on what might have gone wrong!

### Multi-Server

A multi-server install shares content, accounts and some settings between multiple Server Managers.

#### Windows

1. Follow steps 1 to 9 in the 'Single Server Windows' guide above.
2. Copy the assetto-multiserver-manager.exe from the zip file to your server manager install folder
3. Copy the servers.yml from the zip file to your server manager install folder
4. Open servers.yml. Set 'num_servers' to the number of servers you want to install. You can change this later!
5. Run assetto-multiserver-manager.exe by double clicking on it.
6. Navigate to http://127.0.0.1:8772.
7. Follow steps 11-18 in the guide above. You should only need to upload content once, and it will be shared between 
   servers. Server Options will need to be configured for each server. You can switch between servers by using the 
   'Multiserver' icon in the top right hand corner of Server Manager.
8. When you want to launch Server Manager, make sure to use assetto-multiserver-manager.exe, not server-manager.exe.

#### Linux

1. Follow steps 1 to 9 in the 'Single Server Linux' guide above.
2. Copy the assetto-multiserver-manager binary from the zip file to your server manager install folder
3. Copy the servers.yml from the zip file to your server manager install folder
4. Open servers.yml. Set 'num_servers' to the number of servers you want to install. You can change this later!
5. Run assetto-multiserver-manager from the terminal, e.g. ./assetto-multiserver-manager
6. Navigate to http://127.0.0.1:8772.
7. Follow steps 11-18 in the guide above. You should only need to upload content once, and it will be shared between 
   servers. Server Options will need to be configured for each server. You can switch between servers by using the 
   'Multiserver' icon in the top right hand corner of Server Manager.
8. When you want to launch Server Manager, make sure to use the assetto-multiserver-manager binary, not server-manager.


### Docker Installation

This guide is brief, and expects that the reader is experienced with Docker.

In the linux zip download of Server Manager (available from emperorservers.com), there is a Dockerfile and a docker-compose.yml file. 

CHANGELOG.txt
v2.3.7
------

Added:

* Added `Practice Entry List Type` option to the Championship form, this option controls the Entry List behaviour for only looping practice events started from the Championship events.
* Added the ability to delete Live Timings times for specific driver/car combinations, rather than only being able to delete all times for each driver.
* Added `TrackInfo` to the available variables for `Discord Notification Message` and `Discord Notification Title`, this will include the track name and a download link if one has been configured.
* Added `View Details Weather` and `View Details Dynamic Track` permissions, which control whether a user can see the weather plan and track grip settings when viewing an event details popup.
* The Championship Overview tab will now state `DQ` instead of `DNS` if a driver has no points for an event because they were disqualified.

Fixed:

* Fixed an issue where fixed setups could not be applied to cars with capitalised letters in their name in some scenarios.
* Fixed an issue where Race Weekend Session forms could be impossible to save, when the Wait Time value was too low and hidden.
* Fixed an issue where the "Start after Parent Session has completed" Race Weekend option could cause Server Manager to crash in some rare circumstances.
* Fixed an issue where the "Duplicate" function on the Race Weekend page would not work properly when the Race Weekend was part of a Championship.
* Improved display of driver standings cards (with car/livery images) in Championships when some drivers have teams and some do not.
* Removed limits on wind speed (enforced by ACSM, limits may still be enforced by the game or CSP).

---

v2.3.6
------

Added:

* Added Delta (to personal best and overall best) to Live Timings.
* Columns in the Connected Driver table on the Live Timings page can now be shown/hidden from the Settings menu on the page. Settings are saved for each individual in browser storage, as is the setting for "Show Collision Popups".
* Added support for Partially Locked Entry Lists in Championships.
* Added a `Zoom` toggle button to the Live Timings map, which will re-arrange the Live Timings page to make the map (and table) larger. To make the most of this feature please rebuild your track maps from the Maintenance section of the Server Options page.
* Added "Fixed Tyre Setups" to the Fixed Setup dropdown. These setups are created automatically by Server Manager and only lock the tyre in the setup. Please note that drivers can still change to other tyre compounds via a pit stop.
* Added a warning to the car upload page if a car without a `data.acd` file is selected.
* Added the ability to download and delete results and drift results from their respective list pages.

Fixed:

* Track URL on Track Details page now correctly links to the provided URL.
* Fixed an issue where a driver that was deleted from Live Timings may re-appear after advancing to the next looped session.
* Fixed an issue where it may not be possible to edit pit lane detection on tracks with multiple layouts.
* The `Use MPH` option is now applied to Speed Trap hits in results files.
* Fixed an issue where the Live Timings may not display correctly for some timing data.
* Fixed an issue where DRS penalties may not activate properly when moving into a race from a different session.
* Improved the accuracy of the DRS detection point used for DRS penalties.
* Fixed an issue where fastest lap points may not be applied for some Race Weekend Events.
* Improved the behaviour of the Auto Loop system. Events will now loop by themselves without requiring a force stop timer if there are no drivers connected and the session ends, with the exception of races with a number of laps defined instead of time.
* Made track temperatures in Live Weather mode less aggressive, also improved accuracy of track temperatures after sunset.
* Improved Live Timings appearance on smaller mobile devices.
* Discord notifications will now show the override password for an event in all contexts, if an override password is set.
* Fixed an issue where events may not save properly if some track files were missing.
* Enforced a minimum value of 1 for `CSP Weather Transition Duration`, to prevent weather conditions from breaking if was set to 0.
* Improved description for `Minimum Number Of Driver Swaps` and `Not Enough Driver Swaps Penalty` to make it more clear that the minimum number of swaps is applied to the group of drivers, not each individual driver.
* Made Race session wait time behaviour more clear in Race Weekend Race sessions. If a warm up session is enabled the minimum wait time is 30 seconds, otherwise the minimum wait time is 120 seconds.
* Fixed an issue where cars that were uploaded together could have kn5 checksums from each other added when one car's name contained the entire other car's name (for example "ks_bmw_m4" and "ks_bmw_m4_akrapovic").

---

v2.3.5
------

Added:

* Added `Second Race` to Weather `Sessions` input, so weather presets can be applied specifically to the second race when `Reversed Grid Race Positions` is set to a non-zero value.

plugins/events.lua
json = require "json"
utils = require "utils"

-- Insert json folder path here, required for getTrackInfo
local jsonPath = "./plugins" -- default should be correct if countryCodes.json is inside plugins folder
local contentPath = "./assetto/content" -- depends on assetto server location

-- these are lua hooks related to the manager itself, for help please view lua_readme.md!
-- there are some example functions here to give you an idea of what is possible, feel free to write your own!
-- if you do and think other people would be interested in them consider making a pull request at https://github.com/JustaPenguin/assetto-server-manager

-- called when any session ends
function onEndSession(encodedResultsFolderAbsolutePath, encodedResultsFileName)
    -- Decode block, you probably shouldn't touch these!
    local resultsFolderAbsolutePath = json.decode(encodedResultsFolderAbsolutePath)
    local resultsFileName = json.decode(encodedResultsFileName)

    -- Function block NOTE: this hook doesn't block, so it can run constantly alongside the manager if required

    -- Open the event results file
    --local encodedResultsJson = utils.jsonOpen(resultsFolderAbsolutePath, resultsFileName)

    -- Decode the event results file
    --success = pcall(function()
    --    resultsJson = json.decode(encodedResultsJson)
    --end)

    --if not success then
    --    print("events.lua: Couldn't decode results file: ", resultsFolderAbsolutePath .. resultsFileName)
    --    return nil
    --end

    -- Print out the results structure and data
    --print("Results:", utils.dump(resultsJson))
end

-- called when any event (including championships/race weekends) is started (for championships this is called AFTER onChampionshipEventStart)
function onEventStart(encodedRaceConfig, encodedServerOpts, encodedEntryList)
    -- Decode block, you probably shouldn't touch these!
    local raceConfig = json.decode(encodedRaceConfig)
    local serverOpts = json.decode(encodedServerOpts)
    local entryList = json.decode(encodedEntryList)

    -- Uncomment these lines and run the function (start any event) to print out the structure of each object.
    --print("Race Config:", utils.dump(raceConfig))
    --print("Server Options:", utils.dump(serverOpts))
    --print("Entry List:", utils.dump(entryList))

    -- Function block NOTE: this hook BLOCKS, make sure your functions don't loop forever!

    -- Uncomment this line to set Weather API On
    -- in order to use the weatherAPI you need to get a free API key from https://openweathermap.org/
    -- NOTE: ACSM now implements Live Weather directly, so this lua plugin is defunct. It will be left here as an example.
    -- raceConfig, serverOpts = getWeatherForTrack(raceConfig, serverOpts, "get-an-api-key-from-https://openweathermap.org/")

    -- Encode block, you probably shouldn't touch these either!
    return json.encode(entryList), json.encode(serverOpts), json.encode(raceConfig)
end

-- called when any NON CHAMPIONSHIP/RACE WEEKEND event is scheduled
function onEventSchedule(encodedRace)
    -- Decode block, you probably shouldn't touch these!
    local race = json.decode(encodedRace)

    -- Uncomment these lines and run the function (start any event) to print out the structure of each object.
    --print("Race:", utils.dump(race))

    -- Function block NOTE: this hook BLOCKS, make sure your functions don't loop forever!


    -- Encode block, you probably shouldn't touch these either!
    return json.encode(race)
end

-- called when any RACE WEEKEND event is scheduled
function onRaceWeekendEventSchedule(encodedRaceWeekendSession, encodedRaceWeekend)
    -- Decode block, you probably shouldn't touch these!
    local session = json.decode(encodedRaceWeekendSession)
    local raceWeekend = json.decode(encodedRaceWeekend)

    -- Uncomment these lines and run the function (start any event) to print out the structure of each object.
    --print("Race Weekend Session:", utils.dump(session))
    --print("Race Weekend:", utils.dump(raceWeekend))

    -- Function block NOTE: this hook BLOCKS, make sure your functions don't loop forever!

    -- Encode block, you probably shouldn't touch these either!
    return json.encode(raceWeekend), json.encode(session)
end

function getWeatherForTrack(raceConfig, serverOpts, apiKey)
    if apiKey == "get-an-api-key-from-https://openweathermap.org/" then
        print("events.lua: No API key is set, could not activate weather API")

        return raceConfig, serverOpts
    end

    -- Getting location from track_ui.json, make sure to set jsonPath and contentPath
    location = getTrackInfo(raceConfig)
    -- if you want to manually set the location for tracks without location info uncomment this line and set the location, you can download a city list here: http://bulk.openweathermap.org/sample/
    -- location = "Manchester,uk"

    if location == nil then
        --Set location manually if no trackinfo is found
        location = "Manchester,UK"
    end

    return weatherAPI(raceConfig, serverOpts, apiKey)
end

-- get track location from his ui.json file, set dynamic weather on if a location is found
-- if not, use weather in custom race config
function getTrackInfo(raceConfig)
    local track = raceConfig["Track"]
    local layout = raceConfig["TrackLayout"]

    local trackPath = contentPath .. "/tracks/" .. track .. "/ui/" .. layout
    local encodedTrackJson = utils.jsonOpen(trackPath, "ui_track.json")
    local trackJson

    success = pcall(function()
        trackJson = json.decode(encodedTrackJson)
    end)

    if not success then
        print("events.lua: Couldn't decode track UI file: ", trackPath .. "/ui_track.json. Falling back to manual setting")
        return nil
    end

    countryFull = trackJson["country"]
    city = trackJson["city"]

    local encodedCountryCodes = utils.jsonOpen(jsonPath, "countryCodes.json")
    local countryCodes = json.decode(encodedCountryCodes)

    if city == nil or countryFull == nil then
        print("events.lua: No country/city found in track UI file, falling back to manual setting")
        return nil
    end

    if countryCodes[countryFull] == nil then
        print("events.lua: Location was found in track UI file (" .. city .. ", " .. countryFull .. "). But was not found in countryCodes.json, please update! Falling back to manual setting")
        return nil
    end

    location = city .. "," .. countryCodes[countryFull]

    return location
end

-- weather API
function weatherAPI(raceConfig, serverOpts, apiKey)
    -- set the weather based on the current weather at 'location'
    local body, status = httpRequest("http://api.openweathermap.org/data/2.5/weather?q=" .. location .. "&APPID=" .. apiKey, "GET", "")

    -- If location not found in openWeatherMap, stop the function and use weather configured in web manager
    if status >= 400 then
        print("events.lua: Weather API HTTP request returned status ", status, ". Weather API deactivated")
        return raceConfig, serverOpts
    end

    local weatherData = json.decode(body)

    -- wind speed, from m/s to km/h
    raceConfig["WindBaseSpeedMin"] = math.floor(weatherData["wind"]["speed"] * 3.6)
    raceConfig["WindBaseSpeedMax"] = raceConfig["WindBaseSpeedMin"] + 2
    -- wind angle
    raceConfig["WindBaseDirection"] = weatherData["wind"]["deg"]
    raceConfig["WindVariationDirection"] = 5
    -- csp rain (will be enabled later if necessary), requires at least CSP 0.1.74-preview3
    raceConfig["CSPRainEnabled"] = false

    -- there should only be one weather, but we'll apply to all just in case
    for name, weather in pairs(raceConfig["Weather"]) do

        -- ambient temp, from Kelvin to Degrees Celcius
        weather["BaseTemperatureAmbient"] = math.floor(weatherData["main"]["temp"] - 273)
        weather["VariationAmbient"] = 0

        -- road temp
        if weatherData["dt"] > weatherData["sys"]["sunrise"] and weatherData["dt"] < weatherData["sys"]["sunset"] then
            -- the sun is up, base road temp should be a bit higher than ambient
            weather["BaseTemperatureRoad"] = 4
        else
            -- sun is down, base road temp is lower than ambient (large assumption, definitely wrong, please improve!)
            weather["BaseTemperatureRoad"] = 0
        end

        weather["VariationRoad"] = 1

        -- weather codes can be found here: https://openweathermap.org/weather-conditions
        local w = weatherData["weather"][1]["id"]

        if raceConfig["IsSol"] == 1 then
            -- with sol (recommended)

            -- set time of day
            weather["CMWFUseCustomTime"] = 1
            weather["CMWFXTime"] = 0
            weather["CMWFXUseCustomDate"] = 1
            weather["CMWFXDateUnModified"] = weatherData["dt"]
            weather["CMWFXDate"] = weatherData["dt"] - (3600 * 5 * weather["CMWFXTimeMulti"]) -- don't ask

            -- force time 5 hour before sunset (18000) to prevent night
            --weather["CMWFXDate"] = (weatherData["sys"]["sunset"] - 18000) + (weatherData["timezone"]) - (3600 * 5 * weather["CMWFXTimeMulti"]) -- don't ask

            -- set graphics (comment this and uncomment the block bellow for no rain)
            if w == 800 then
                weather["CMGraphics"] = "sol_01_CLear";
                weather["CMWFXType"] = 15;
            elseif w == 801 then
                weather["CMGraphics"] = "sol_02_Few Clouds";
                weather["CMWFXType"] = 16
            elseif w == 802 then
                weather["CMGraphics"] = "sol_03_Scattered Clouds";
                weather["CMWFXType"] = 17
                --Do not uncomment elseif w ==  then weather["CMGraphics"] = "sol_04_Windy"; weather["CMWFXType"] = 31 --no real weather for windy
            elseif w == 803 then
                weather["CMGraphics"] = "sol_05_Broken Clouds";
                weather["CMWFXType"] = 18
            elseif w == 804 then
                weather["CMGraphics"] = "sol_06_Overcast";
                weather["CMWFXType"] = 19
            elseif w == 701 then
                weather["CMGraphics"] = "sol_11_Mist";
                weather["CMWFXType"] = 21
            elseif w == 741 then
                weather["CMGraphics"] = "sol_12_Fog";
                weather["CMWFXType"] = 20
            elseif w == 721 then
                weather["CMGraphics"] = "sol_21_Haze";
                weather["CMWFXType"] = 23
            elseif w == 731 then
                weather["CMGraphics"] = "sol_22_Dust";
                weather["CMWFXType"] = 25
            elseif w == 751 then
                weather["CMGraphics"] = "sol_23_Sand";
                weather["CMWFXType"] = 24
            elseif w == 711 then
                weather["CMGraphics"] = "sol_24_Smoke";
                weather["CMWFXType"] = 22
            elseif w == 300 then
                weather["CMGraphics"] = "sol_31_Light Drizzle";
                weather["CMWFXType"] = 3

                -- csp rain, you may want to tweak intensity
                raceConfig["CSPRainEnabled"] = true
                raceConfig["CSPRainIntensity"] = 10
            elseif w == 301 then
                weather["CMGraphics"] = "sol_32_Drizzle";
                weather["CMWFXType"] = 4

                raceConfig["CSPRainEnabled"] = true
                raceConfig["CSPRainIntensity"] = 20
            elseif w >= 302 and w <= 321 then
                weather["CMGraphics"] = "sol_33_Heavy Drizzle";
                weather["CMWFXType"] = 5

                raceConfig["CSPRainEnabled"] = true
                raceConfig["CSPRainIntensity"] = 25
            elseif w == 500 then
                weather["CMGraphics"] = "sol_34_Light Rain";
                weather["CMWFXType"] = 6

                raceConfig["CSPRainEnabled"] = true
                raceConfig["CSPRainIntensity"] = 40
            elseif w == 501 then
                weather["CMGraphics"] = "sol_35_Rain";
                weather["CMWFXType"] = 7

                raceConfig["CSPRainEnabled"] = true
                raceConfig["CSPRainIntensity"] = 50
            elseif w >= 502 and w <= 531 then
                weather["CMGraphics"] = "sol_36_Heavy Rain";
                weather["CMWFXType"] = 8

                raceConfig["CSPRainEnabled"] = true
                raceConfig["CSPRainIntensity"] = 60
            elseif w == 200 or w == 210 or w == 230 then
                weather["CMGraphics"] = "sol_41_Light Thunderstorm";
                weather["CMWFXType"] = 0

                raceConfig["CSPRainEnabled"] = true
                raceConfig["CSPRainIntensity"] = 75
            elseif w == 201 or w == 211 or w == 231 then
                weather["CMGraphics"] = "sol_42_Thunderstorm";
                weather["CMWFXType"] = 1

                raceConfig["CSPRainEnabled"] = true
                raceConfig["CSPRainIntensity"] = 85
            elseif w == 202 or w == 212 or w == 221 or w == 232 then
                weather["CMGraphics"] = "sol_43_Heavy Thunderstorm";
                weather["CMWFXType"] = 2

                raceConfig["CSPRainEnabled"] = true
                raceConfig["CSPRainIntensity"] = 95
            elseif w == 771 then
                weather["CMGraphics"] = "sol_44_Squalls";
                weather["CMWFXType"] = 26
            elseif w == 781 then
                weather["CMGraphics"] = "sol_45_Tornado";
                weather["CMWFXType"] = 27
                --Do not uncomment elseif w ==  then weather["CMGraphics"] = "sol_46_Hurricane"; weather["CMWFXType"] = 28 --no real weather for hurricane
            elseif w == 600 or w == 620 then
                weather["CMGraphics"] = "sol_51_Light Snow";
                weather["CMWFXType"] = 9
            elseif w == 601 or w == 621 then
                weather["CMGraphics"] = "sol_52_Snow";
                weather["CMWFXType"] = 10
            elseif w == 602 or w == 622 then
                weather["CMGraphics"] = "sol_53_Heavy Snow";
                weather["CMWFXType"] = 11
            elseif w == 611 or w == 615 then
                weather["CMGraphics"] = "sol_54_Light Sleet";
                weather["CMWFXType"] = 12
            elseif w == 612 or w == 616 then
                weather["CMGraphics"] = "sol_55_Sleet";
                weather["CMWFXType"] = 13
            elseif w == 613 then
                weather["CMGraphics"] = "sol_56_Heavy Sleet";
                weather["CMWFXType"] = 14
                --Do not uncomment elseif w ==  then weather["CMGraphics"] = "sol_57_Hail"; weather["CMWFXType"] = 32 --no real weather for hail
            end

            -- set graphics no rain (comment the block above and uncomment this one for no rain)
            --if     w == 800 then weather["CMGraphics"] = "sol_01_CLear"; weather["CMWFXType"] = 15;
            --elseif w == 801 then weather["CMGraphics"] = "sol_02_Few Clouds"; weather["CMWFXType"] = 16
            --elseif w == 802 then weather["CMGraphics"] = "sol_03_Scattered Clouds"; weather["CMWFXType"] = 17
            --Do not uncomment elseif w ==  then weather["CMGraphics"] = "sol_04_Windy"; weather["CMWFXType"] = 31 --no real weather for windy
            --elseif w == 803 then weather["CMGraphics"] = "sol_05_Broken Clouds"; weather["CMWFXType"] = 18
            --elseif w == 804 then weather["CMGraphics"] = "sol_06_Overcast"; weather["CMWFXType"] = 19
            --elseif w == 701 then weather["CMGraphics"] = "sol_11_Mist"; weather["CMWFXType"] = 21
            --elseif w == 741 then weather["CMGraphics"] = "sol_12_Fog"; weather["CMWFXType"] = 20
            --elseif w == 721 then weather["CMGraphics"] = "sol_21_Haze"; weather["CMWFXType"] = 23
            --elseif w == 731 then weather["CMGraphics"] = "sol_22_Dust"; weather["CMWFXType"] = 25
            --elseif w == 751 then weather["CMGraphics"] = "sol_23_Sand"; weather["CMWFXType"] = 24
            --elseif w == 711 then weather["CMGraphics"] = "sol_24_Smoke"; weather["CMWFXType"] = 22
            --elseif w == 300 then weather["CMGraphics"] = "sol_05_Broken Clouds"; weather["CMWFXType"] = 18
            --elseif w == 301 then weather["CMGraphics"] = "sol_06_Overcast"; weather["CMWFXType"] = 19
            --elseif w >= 302 and w <= 321 then weather["CMGraphics"] = "sol_24_Smoke"; weather["CMWFXType"] = 22
            --elseif w == 500 then weather["CMGraphics"] = "sol_05_Broken Clouds"; weather["CMWFXType"] = 18
            --elseif w == 501 then weather["CMGraphics"] = "sol_06_Overcast"; weather["CMWFXType"] = 19
            --elseif w >= 502 and w <= 531 then weather["CMGraphics"] = "sol_24_Smoke"; weather["CMWFXType"] = 22
            --elseif w == 200 or w == 210 or w == 230 then weather["CMGraphics"] = "sol_06_Overcast"; weather["CMWFXType"] = 19
            --elseif w == 201 or w == 211 or w == 231 then weather["CMGraphics"] = "sol_24_Smoke"; weather["CMWFXType"] = 22
            --elseif w == 202 or w == 212 or w == 221 or w == 232 then weather["CMGraphics"] = "sol_24_Smoke"; weather["CMWFXType"] = 22
            --elseif w == 771 then weather["CMGraphics"] = "sol_44_Squalls"; weather["CMWFXType"] = 26
            --elseif w == 781 then weather["CMGraphics"] = "sol_24_Smoke"; weather["CMWFXType"] = 22
            --Do not uncomment elseif w ==  then weather["CMGraphics"] = "sol_46_Hurricane"; weather["CMWFXType"] = 28 --no real weather for hurricane
            --elseif w == 600 or w == 620 then weather["CMGraphics"] = "sol_05_Broken Clouds"; weather["CMWFXType"] = 18
            --elseif w == 601 or w == 621 then weather["CMGraphics"] = "sol_12_Fog"; weather["CMWFXType"] = 20
            --elseif w == 602 or w == 622 then weather["CMGraphics"] = "sol_24_Smoke"; weather["CMWFXType"] = 22
            --elseif w == 611 or w == 615 then weather["CMGraphics"] = "sol_05_Broken Clouds"; weather["CMWFXType"] = 18
            --elseif w == 612 or w == 616 then weather["CMGraphics"] = "sol_12_Fog"; weather["CMWFXType"] = 20
            --elseif w == 613 then weather["CMGraphics"] = "sol_06_Overcast"; weather["CMWFXType"] = 19
            --Do not uncomment elseif w ==  then weather["CMGraphics"] = "sol_57_Hail"; weather["CMWFXType"] = 32 --no real weather for hail
            --end

            weather["Graphics"] = weather["CMGraphics"] .. "_type=" .. weather["CMWFXType"] .. "_time=0_mult=" .. weather["CMWFXTimeMulti"] .. "_start=" .. weather["CMWFXDate"]
        else
            -- without Sol (not recommended)
            -- you could set sun angle from time of day here, I'm not going to though (just use Sol)

            -- set graphics
            if w == 800 then
                weather["Graphics"] = "3_clear"
            elseif w == 801 then
                weather["Graphics"] = "4_mid_clear"
            elseif w == 802 then
                weather["Graphics"] = "5_light_clouds"
            elseif w == 803 then
                weather["Graphics"] = "6_mid_clouds"
            elseif w == 804 then
                weather["Graphics"] = "7_heavy_clouds"
            elseif w == 741 then
                weather["Graphics"] = "2_light_fog"
            end
        end

    end

    -- Add text to server name to indicate this has been done
    serverOpts["Name"] = serverOpts["Name"] .. " | Weather Live From " .. weatherData["name"]

    return raceConfig, serverOpts
end

--{ Response json from openweathermap looks like this
--	"coord": {
--		"lon": -0.13,
--		"lat": 51.51
--	},
--	"weather": [{
--		"id": 803,
--		"main": "Clouds",
--		"description": "broken clouds",
--		"icon": "04d"
--	}],
--	"base": "stations",
--	"main": {
--		"temp": 281.13,
--		"pressure": 998,
--		"humidity": 66,
--		"temp_min": 280.15,
--		"temp_max": 282.15
--	},
--	"visibility": 10000,
--	"wind": {
--		"speed": 3.1,
--		"deg": 240
--	},
--	"rain": {},
--	"clouds": {
--		"all": 67
--	},
--	"dt": 1573656582,
--	"sys": {
--		"type": 1,
--		"id": 1414,
--		"country": "GB",
--		"sunrise": 1573629270,
--		"sunset":  1573661710
--	},
--	"timezone": 0,
--	"id": 2643743,
--	"name": "London",
--	"cod": 200
--}
